### PR TITLE
enable Mixtral-8x7B

### PR DESCRIPTION
**Pull request by jychen-habana**
# What does this PR do?

As title, enable Mixtral-8x7B model

**Test with BS=4:**
**input 1: ('DeepSpeed is a machine learning framework',)**
output 1: ('DeepSpeed is a machine learning framework that enables training of large models on a single machine with multiple GPUs. It is designed to be easy to use and efficient, and it supports a wide range of models and tasks.\n\nDeepSpeed is a deep learning framework that enables training of large models on a single machine with multiple GPUs. It is designed to be easy to use and efficient, and it supports a wide range of models and tasks.\n\nDeepSpeed is a deep learning framework that enables training of large models on a single machine with multiple GPUs. It is designed to be easy to use and efficient, and it supports a wide range of models and tasks',)
**input 2: ('He is working on',)**
output 1: ("He is working on a new album, which is expected to be released in 2019.\n\n## Some Related Sentences\n\nHe and is\namazingly light on his feet for a man his size.\n`` not going to be in the suite when the shooting starts ''.\nsaid he'd take care of the two of us.\nsaid he'd take care of us.\nsaid he'd take care of the two of us ''.\nsaid he'd take care of us ''.\nsaid he'd take care of the two of us ''.\nsaid",)
**input 3: ('He has a',)**
output 1: ('He has a new book out, and he’s on a book tour.\n\nHe’s a former president of the United States, and he’s a former president of the United States.\n\nHe’s a former president, and he’s a former president.\n\nHe’s a former president.\n\nHe’s a former president.\n\nHe’s a former president.\n\nHe’s a former president.\n\nHe’s a former president.\n\nHe’s a former president.\n\nHe’s a former president.\n\nHe’s a former president.\n',)
**input 4: ('He got all',)**
output 1: ('He got all the way to the top of the mountain, but he couldn’t see the view.\n\nThe man had climbed the mountain for days, and he was exhausted. He had been looking forward to seeing the view from the top, but when he got there, he was disappointed. The view was obscured by clouds.\n\nHe sat down on a rock and took a deep breath. He had come so far, and he was so close to his goal. But he couldn’t help but feel a little bit disappointed.\n\nHe looked around, and he saw that there were other people on the mountain. Some of them',)
**Stats:**
Throughput (including tokenization) = **93.95431200903332** tokens/second
Number of HPU graphs                = 22
Memory allocated                    = 87.92 GB
Max memory allocated                = 88.22 GB
Total memory available              = 94.62 GB
Graph compilation duration          = 17.076973195013124 seconds

**Test with FP8 and BS=4:**
**input 1: ('DeepSpeed is a machine learning framework',)**
output 1: ('DeepSpeed is a machine learning framework that enables training of large models on a single machine with a single GPU. It is designed to be easy to use and efficient, and it can be used to train models on a variety of tasks.\n\n## Introduction\n\nDeepSpeed is a machine learning framework that enables training of large models on a single machine with a single GPU. It is designed to be easy to use and efficient, and it can be used to train models on a variety of tasks.\n\n## What is DeepSpeed?\n\nDeepSpeed is a machine learning framework that enables training of large models on a single machine with a single GPU. It is designed',)
**input 2: ('He is working on',)**
output 1: ("He is working on a new album, which is expected to be released in 2019.\n\n## Some Related Sentences\n\nHe and is\namazingly light on his feet for a man his size.\n`` a very stubborn man '', she said.\nsaid he'd call back but he never did.\na very stubborn man '', she said.\nsaid he'd call back but he never did.\nsaid he'd call back but he never did.\nsaid he'd call back but he never did.\nsaid he'd call back but he never did",)
**input 3: ('He has a',)**
output 1: ('He has a new book out, and he’s on a book tour.\n\nHe’s a former president of the United States, and he’s a former president of the United States.\n\nHe’s a former president, and he’s a former president.\n\nHe’s a former president.\n\nHe’s a former president.\n\nHe’s a former president.\n\nHe’s a former president.\n\nHe’s a former president.\n\nHe’s a former president.\n\nHe’s a former president.\n\nHe’s a former president.\n',)
**input 4: ('He got all',)**
output 1: ('He got all the way to the end of the interview and then he asked me, “What’s your biggest weakness?”\n\nI was stumped. I had been asked this question before, but I had never really thought about it. I had never really thought about what my biggest weakness was.\n\nI had never really thought about what my biggest weakness was.\n\nI had never really thought about what my biggest weakness was.\n\nI had never really thought about what my biggest weakness was.\n\nI had never really thought about what my biggest weakness was.\n\nI had never really thought about what my biggest weakness was.',)
**Stats:**
Throughput (including tokenization) = **156.10840403691122** tokens/second
Number of HPU graphs                = 19
Memory allocated                    = 45.88 GB
Max memory allocated                = 46.13 GB
Total memory available              = 94.62 GB
Graph compilation duration          = 262.93332553398795 seconds

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by jychen-habana** [comment]
The model is ready, code needs further improvement before merge.

**Comment by regisss** [comment]
Is it supposed to work with DeepSpeed on several devices?

**Comment by jychen-habana** [comment]
> Is it supposed to work with DeepSpeed on several devices?

Yes, for TP the modeling script is ready, but something also needs to change in Habana DeepSpeed auto_tp.

File: optimum/habana/transformers/models/mixtral/modeling_mixtral.py [hunk 1]
```diff
@@ -0,0 +1,690 @@
...
+        return FusedRoPE.apply(
+            q, cos.unsqueeze(0).unsqueeze(0), sin.unsqueeze(0).unsqueeze(0), position_ids
+        ), FusedRoPE.apply(k, cos.unsqueeze(0).unsqueeze(0), sin.unsqueeze(0).unsqueeze(0), position_ids)
+    else:
+        return apply_rotary_pos_emb(q, k, cos, sin, position_ids)
+
+
+def gaudi_mixtral_rmsnorm_forward(self, hidden_states):
+    """
+    Copied from MixtralRMSNorm.forward: https://github.com/huggingface/transformers/blob/main/src/transformers/models/mixtral/modeling_mixtral.py
```
Comment (ZhaiFeiyue) [inline optimum/habana/transformers/models/mixtral/modeling_mixtral.py:102]:
```suggestion
    Copied from MixtralRMSNorm.forward: https://github.com/huggingface/transformers/blob/v4.37.0/src/transformers/models/mixtral/modeling_mixtral.py
```

File: optimum/habana/transformers/models/mixtral/modeling_mixtral.py [hunk 2]
```diff
@@ -0,0 +1,690 @@
...
+
+def gaudi_mixtral_repeat_kv(
+    query_states: torch.Tensor,
+    key_states: torch.Tensor,
+    value_states: torch.Tensor,
+    attention_mask: torch.Tensor,
+    n_rep: int,
+):
+    """
+    Copied from repeat_kv: https://github.com/huggingface/transformers/blob/main/src/transformers/models/mixtral/modeling_mixtral.py
```
Comment (ZhaiFeiyue) [inline optimum/habana/transformers/models/mixtral/modeling_mixtral.py:131]:
```suggestion
    Copied from repeat_kv: https://github.com/huggingface/transformers/blob/v4.37.0/src/transformers/models/mixtral/modeling_mixtral.py
```

File: optimum/habana/transformers/models/mixtral/modeling_mixtral.py [hunk 3]
```diff
@@ -0,0 +1,690 @@
...
+    inputs_embeds: Optional[torch.FloatTensor] = None,
+    use_cache: Optional[bool] = None,
+    output_attentions: Optional[bool] = None,
+    output_hidden_states: Optional[bool] = None,
+    output_router_logits: Optional[bool] = None,
+    return_dict: Optional[bool] = None,
+    token_idx: Optional[torch.Tensor] = None,
+) -> Union[Tuple, MoeModelOutputWithPast]:
+    """
+    Copied from LlamaModel.forward: https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py
```
Comment (ZhaiFeiyue) [inline optimum/habana/transformers/models/mixtral/modeling_mixtral.py:407]:
```suggestion
    Copied from MixtralModel.forward: https://github.com/huggingface/transformers/blob/v4.37.0/src/transformers/models/mixtral/modeling_mixtral.py#L1069
```

File: optimum/habana/transformers/models/mixtral/modeling_mixtral.py [hunk 4]
```diff
@@ -0,0 +1,690 @@
...
+        past_key_values=next_cache,
+        hidden_states=all_hidden_states,
+        attentions=all_self_attns,
+        router_logits=all_router_logits,
+    )
+
+
+class GaudiMixtralForCausalLM(MixtralForCausalLM):
+    """
+    Inherits from MixtralForCausalLM: https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py
```
Comment (ZhaiFeiyue) [inline optimum/habana/transformers/models/mixtral/modeling_mixtral.py:564]:
```suggestion
    Inherits from MixtralForCausalLM: https://github.com/huggingface/transformers/blob/v4.37.0/src/transformers/models/mixtral/modeling_mixtral.py#L1231
```

File: optimum/habana/transformers/modeling_utils.py [hunk 5]
```diff
@@ -95,6 +96,11 @@
     gaudi_mistral_attention_forward,
     gaudi_mistral_decoder_layer_forward,
     gaudi_mistral_model_forward,
+    gaudi_mixtral_attn_forward,
```
Comment (yafshar) [inline optimum/habana/transformers/modeling_utils.py:99]:
```suggestion
    gaudi_mixtral_attention_forward,
```

File: optimum/habana/transformers/models/mixtral/modeling_mixtral.py [hunk 6]
```diff
@@ -0,0 +1,690 @@
...
+
+    attn_output = self.o_proj(attn_output)
+
+    if not output_attentions:
+        attn_weights = None
+
+    return attn_output, attn_weights, past_key_value
+
+
+def gaudi_mixtral_block_sparse_moe_forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
```
Comment (yafshar) [inline optimum/habana/transformers/models/mixtral/modeling_mixtral.py:285]:
```suggestion
def gaudi_mixtral_block_sparse_moe_forward(self, hidden_states: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
```

**Comment by jychen-habana** [comment]
> @jychen-habana thanks for this PR. I think we should close #716 Do you know when the change in Habana DeepSpeed are available? I do not want to push some changes there and duplicate the work.

We are submitting a PR for DeepSpeed. For the modeling script, the changes for TP are already included.

File: optimum/habana/transformers/models/mixtral/modeling_mixtral.py [hunk 7]
```diff
@@ -0,0 +1,690 @@
...
+    past_key_values_length = 0
+
+    if self.gradient_checkpointing and self.training:
+        if use_cache:
+            logger.warning_once(
+                "`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`..."
+            )
+            use_cache = False
+
+    # use_cache = False
```
Comment (regisss) [inline optimum/habana/transformers/models/mixtral/modeling_mixtral.py:441]: To remove

File: optimum/habana/transformers/models/mixtral/modeling_mixtral.py [hunk 8]
```diff
@@ -0,0 +1,690 @@
...
+            attentions=outputs.attentions,
+            router_logits=outputs.router_logits,
+        )
+
+    def prepare_inputs_for_generation(
+        self, input_ids, past_key_values=None, attention_mask=None, inputs_embeds=None, **kwargs
+    ):
+        token_idx = kwargs.get("token_idx", None)
+
+        if past_key_values:
```
Comment (regisss) [inline optimum/habana/transformers/models/mixtral/modeling_mixtral.py:658]:
```suggestion
        if past_key_values is not None:
```

File: optimum/habana/transformers/models/mixtral/modeling_mixtral.py [hunk 9]
```diff
@@ -0,0 +1,690 @@
...
+            if token_idx is None:
+                input_ids = input_ids[:, -1:]
```
Comment (regisss) [inline optimum/habana/transformers/models/mixtral/modeling_mixtral.py:660]: Let's put here the logic of Transformers v4.37: https://github.com/huggingface/transformers/blob/v4.37.2/src/transformers/models/mixtral/modeling_mixtral.py#L1376-L1402

File: optimum/habana/transformers/models/mixtral/modeling_mixtral.py [hunk 10]
```diff
@@ -0,0 +1,690 @@
...
+        position_ids = kwargs.get("position_ids", None)
+        if attention_mask is not None and position_ids is None:
+            # create position_ids on the fly for batch generation
+            position_ids = attention_mask.long().cumsum(-1) - 1
+            position_ids.masked_fill_(attention_mask == 0, 1)
+            if past_key_values:
+                if token_idx is not None:
+                    position_ids = torch.index_select(position_ids, 1, token_idx - 1)
+                else:
+                    position_ids = position_ids[:, -1].unsqueeze(-1)
```
Comment (regisss) [inline optimum/habana/transformers/models/mixtral/modeling_mixtral.py:673]:
```suggestion
                    position_ids = position_ids[:, -input_ids.shape[1] :]
```
from Transformers v4.37: https://github.com/huggingface/transformers/blob/v4.37.2/src/transformers/models/mixtral/modeling_mixtral.py#L1410

File: optimum/habana/transformers/models/mixtral/modeling_mixtral.py [hunk 11]
```diff
@@ -0,0 +1,690 @@
...
+            ), f"inp_seq_len must be the same. self.inp_seq_len:{self.inp_seq_len} inp_seq_len:{inp_seq_len}"
+            self.cache.fill_(0)
+
+    def get_shape(self):
+        if self.cache is None:
+            return None
+        return self.cache.shape
+
+    def forward(self, cur, dim, idx):
+        return update(self.cache, cur, dim, idx, self.inp_seq_len)
```
[the remark this answers is not in the document]
Comment (jychen-habana) [reply optimum/habana/transformers/models/mixtral/modeling_mixtral.py:181]: Yes, they are from Llama

**Comment by yafshar** [comment]
@jychen-habana, I am unable to repeat the similar stats as you
with `use_hpu_graphs`, on 1 device, and with BF16, I am getting 
```sh
>> python run_generation.py --model_name_or_path mistralai/Mixtral-8x7B-v0.1 --use_kv_cache --use_hpu_graphs  --max_new_tokens 100 --bf16

Throughput (including tokenization) = 23.787950727713117 tokens/second
Number of HPU graphs                = 18
Memory allocated                    = 87.66 GB
Max memory allocated                = 87.71 GB
Total memory available              = 94.62 GB
Graph compilation duration          = 13.244057161995443 seconds
```
without it
```sh
>> python run_generation.py --model_name_or_path mistralai/Mixtral-8x7B-v0.1 --use_kv_cache --max_new_tokens 100 --bf16

Throughput (including tokenization) = 14.394814172747322 tokens/second
Number of HPU graphs                = 11
Memory allocated                    = 87.54 GB
Max memory allocated                = 87.61 GB
Total memory available              = 94.62 GB
Graph compilation duration          = 20.91424718502094 seconds
```

**Comment by jychen-habana** [comment]
> @jychen-habana, I am unable to repeat the similar stats as you with `use_hpu_graphs`, on 1 device, and with BF16, I am getting
> 
> ```shell
> >> python run_generation.py --model_name_or_path mistralai/Mixtral-8x7B-v0.1 --use_kv_cache --use_hpu_graphs  --max_new_tokens 100 --bf16
> 
> Throughput (including tokenization) = 23.787950727713117 tokens/second
> Number of HPU graphs                = 18
> Memory allocated                    = 87.66 GB
> Max memory allocated                = 87.71 GB
> Total memory available              = 94.62 GB
> Graph compilation duration          = 13.244057161995443 seconds
> ```
> 
> without it
> 
> ```shell
> >> python run_generation.py --model_name_or_path mistralai/Mixtral-8x7B-v0.1 --use_kv_cache --max_new_tokens 100 --bf16
> 
> Throughput (including tokenization) = 14.394814172747322 tokens/second
> Number of HPU graphs                = 11
> Memory allocated                    = 87.54 GB
> Max memory allocated                = 87.61 GB
> Total memory available              = 94.62 GB
> Graph compilation duration          = 20.91424718502094 seconds
> ```

23.787950727713117 tokens/second for bs=1 is expected; what I tested (93.95431200903332 tokens/second) is for bs=4.

**Comment by jychen-habana** [comment]
@regisss, since the PR to DeepSpeed may not be merged quickly, currently Mixtral-8x7B inference can only be run with a single card, also with FP8 HQT(as described in 'What does this PR do?'). Please help to review and merge.

**Comment by yafshar** [comment]
@regisss, @jychen-habana, I also validated the DeepSpeed PR https://github.com/HabanaAI/DeepSpeed/pull/3 for BF16 on 4 & 8 devices. I see no issues

**Review by regisss — APPROVED** [review]
LGTM!

I also pushed a commit to add a baseline for Mixtral for automatic regression tests.
Besides, I had to run `make style` again due to a new version of `ruff` and, since I can't update the fork's main branch, there are a lot of small changes that were added to this PR. Sorry for this, but that enables me to merge the PR faster :slightly_smiling_face:

**Pull request merged**